[Bug] Suwayomi Database storing old metadata (bato source)

**PLEASE READ THIS**

I acknowledge that:

- I have updated to the latest version of the app.
- I have tried the troubleshooting guide described in `README.md`
- If this is a request for adding/changing an extension it should be brought up to your extension repo.
- If this is an issue with some extension not working properly, It does work inside Tachiyomi as intended.
- I have searched the existing issues and this is a new ticket **NOT** a duplicate or related to another open issue
- I will fill out the title and the information in this template

Note that the issue will be automatically closed if you do not fill out the title or requested information.

---

## Device information
- Suwayomi-Server version: v1.0.0-r1527-docker)
- Server Operating System: docker
- Server Desktop Environment: docker
- Server JVM version: docker
- Client Operating System: docker
- Client Web Browser: Chrome (newest version)

## Steps to reproduce
1. Open bato library
2. Some thumbnails don't show up when browsing the source

### Expected behavior
All thumbnails should be loaded while browsing library

### Actual behavior
Some thumbnails don't load - they error out and even reply doesn't work. To reload a picture you have to open the manga/manhwa with the thumbnail/cover problem and refresh it from inside - then it works correctly
It seems the server tries to connect to a CDN server that doesn't exist anymore:

log:

> 

> 18:08:39.439 [DefaultDispatcher-worker-9] ERROR suwayomi.tachidesk.server.JavalinSetup -- IOException while handling the request
> 
> java.io.IOException: xfs-s117.batcg.org: Name or service not known
> 	at suwayomi.tachidesk.manga.impl.Manga.fetchHttpSourceMangaThumbnail(Manga.kt:291)
> 	at suwayomi.tachidesk.manga.impl.Manga.fetchHttpSourceMangaThumbnail$default(Manga.kt:271)
> 	at suwayomi.tachidesk.manga.impl.Manga$fetchMangaThumbnail$2.invokeSuspend(Manga.kt:320)
> 	at suwayomi.tachidesk.manga.impl.Manga$fetchMangaThumbnail$2.invoke(Manga.kt)
> 	at suwayomi.tachidesk.manga.impl.Manga$fetchMangaThumbnail$2.invoke(Manga.kt)
> 	at suwayomi.tachidesk.manga.impl.util.storage.ImageResponse.getImageResponse(ImageResponse.kt:70)
> 	at suwayomi.tachidesk.manga.impl.Manga.fetchMangaThumbnail(Manga.kt:319)
> 	at suwayomi.tachidesk.manga.impl.Manga.getMangaThumbnail(Manga.kt:365)
> 	at suwayomi.tachidesk.manga.controller.MangaController$thumbnail$2$1.invokeSuspend(MangaController.kt:94)
> 	at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:33)
> 	at kotlinx.coroutines.DispatchedTask.run(DispatchedTask.kt:108)
> 	at kotlinx.coroutines.internal.LimitedDispatcher$Worker.run(LimitedDispatcher.kt:115)
> 	at kotlinx.coroutines.scheduling.TaskImpl.run(Tasks.kt:103)
> 	at kotlinx.coroutines.scheduling.CoroutineScheduler.runSafely(CoroutineScheduler.kt:584)
> 	at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.executeTask(CoroutineScheduler.kt:793)
> 	at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.runWorker(CoroutineScheduler.kt:697)
> 	at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.run(CoroutineScheduler.kt:684)
> Caused by: java.net.UnknownHostException: xfs-s117.batcg.org: Name or service not known
> 	at java.base/java.net.Inet4AddressImpl.lookupAllHostAddr(Native Method)
> 	at java.base/java.net.InetAddress$PlatformNameService.lookupAllHostAddr(Unknown Source)
> 	at java.base/java.net.InetAddress.getAddressesFromNameService(Unknown Source)
> 	at java.base/java.net.InetAddress$NameServiceAddresses.get(Unknown Source)
> 	at java.base/java.net.InetAddress.getAllByName0(Unknown Source)
> 	at java.base/java.net.InetAddress.getAllByName(Unknown Source)
> 	at java.base/java.net.InetAddress.getAllByName(Unknown Source)
> 	at okhttp3.Dns$Companion$DnsSystem.lookup(Dns.kt:49)
> 	at okhttp3.internal.connection.RouteSelector.resetNextInetSocketAddress(RouteSelector.kt:169)
> 	at okhttp3.internal.connection.RouteSelector.nextProxy(RouteSelector.kt:132)
> 	at okhttp3.internal.connection.RouteSelector.next(RouteSelector.kt:74)
> 	at okhttp3.internal.connection.RealRoutePlanner.planConnect(RealRoutePlanner.kt:147)
> 	at okhttp3.internal.connection.RealRoutePlanner.plan(RealRoutePlanner.kt:67)
> 	at okhttp3.internal.connection.FastFallbackExchangeFinder.launchTcpConnect(FastFallbackExchangeFinder.kt:118)
> 	at okhttp3.internal.connection.FastFallbackExchangeFinder.find(FastFallbackExchangeFinder.kt:62)
> 	at okhttp3.internal.connection.RealCall.initExchange$okhttp(RealCall.kt:267)
> 	at okhttp3.internal.connection.ConnectInterceptor.intercept(ConnectInterceptor.kt:32)
> 	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.kt:109)
> 	at okhttp3.internal.cache.CacheInterceptor.intercept(CacheInterceptor.kt:95)
> 	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.kt:109)
> 	at okhttp3.internal.http.BridgeInterceptor.intercept(BridgeInterceptor.kt:84)
> 	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.kt:109)
> 	at okhttp3.internal.http.RetryAndFollowUpInterceptor.intercept(RetryAndFollowUpInterceptor.kt:65)
> 	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.kt:109)
> 	at eu.kanade.tachiyomi.network.interceptor.CloudflareInterceptor.intercept(CloudflareInterceptor.kt:42)
> 	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.kt:109)
> 	at eu.kanade.tachiyomi.network.interceptor.UserAgentInterceptor.intercept(UserAgentInterceptor.kt:19)
> 	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.kt:109)
> 	at eu.kanade.tachiyomi.network.interceptor.UncaughtExceptionInterceptor.intercept(UncaughtExceptionInterceptor.kt:18)
> 	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.kt:109)
> 	at okhttp3.internal.connection.RealCall.getResponseWithInterceptorChain$okhttp(RealCall.kt:205)
> 	at okhttp3.internal.connection.RealCall$AsyncCall.run(RealCall.kt:537)
> 	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(Unknown Source)
> 	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(Unknown Source)
> 	at java.base/java.lang.Thread.run(Unknown Source)



EDIT: The problem is connected with Suwayomi storing old metadata in the database file.
For now the only fix is to backup your manga database and restore it on a fresh database

--- a/server/src/main/kotlin/suwayomi/tachidesk/graphql/mutations/SourceMutation.kt
+++ b/server/src/main/kotlin/suwayomi/tachidesk/graphql/mutations/SourceMutation.kt
@@ -19,7 +19,7 @@ import suwayomi.tachidesk.graphql.types.SourceMetaType
 import suwayomi.tachidesk.graphql.types.SourceType
 import suwayomi.tachidesk.graphql.types.preferenceOf
 import suwayomi.tachidesk.graphql.types.updateFilterList
-import suwayomi.tachidesk.manga.impl.MangaList.insertOrGet
+import suwayomi.tachidesk.manga.impl.MangaList.insertOrUpdate
 import suwayomi.tachidesk.manga.impl.Source
 import suwayomi.tachidesk.manga.impl.util.source.GetCatalogueSource
 import suwayomi.tachidesk.manga.model.table.MangaTable
@@ -135,7 +135,7 @@ class SourceMutation {
                         }
                     }
 
-                val mangaIds = mangasPage.insertOrGet(sourceId)
+                val mangaIds = mangasPage.insertOrUpdate(sourceId)
 
                 val mangas =
                     transaction {

--- a/server/src/main/kotlin/suwayomi/tachidesk/manga/impl/MangaList.kt
+++ b/server/src/main/kotlin/suwayomi/tachidesk/manga/impl/MangaList.kt
@@ -8,14 +8,17 @@ package suwayomi.tachidesk.manga.impl
  * file, You can obtain one at https://mozilla.org/MPL/2.0/. */
 
 import eu.kanade.tachiyomi.source.model.MangasPage
+import org.jetbrains.exposed.dao.id.EntityID
 import org.jetbrains.exposed.sql.and
 import org.jetbrains.exposed.sql.batchInsert
 import org.jetbrains.exposed.sql.select
+import org.jetbrains.exposed.sql.statements.BatchUpdateStatement
 import org.jetbrains.exposed.sql.transactions.transaction
 import suwayomi.tachidesk.manga.impl.util.source.GetCatalogueSource.getCatalogueSourceOrStub
 import suwayomi.tachidesk.manga.model.dataclass.PagedMangaListDataClass
 import suwayomi.tachidesk.manga.model.table.MangaTable
 import suwayomi.tachidesk.manga.model.table.toDataClass
+import java.time.Instant
 
 object MangaList {
     fun proxyThumbnailUrl(mangaId: Int): String {
@@ -44,12 +47,12 @@ object MangaList {
         return mangasPage.processEntries(sourceId)
     }
 
-    fun MangasPage.insertOrGet(sourceId: Long): List<Int> {
+    fun MangasPage.insertOrUpdate(sourceId: Long): List<Int> {
         return transaction {
             val existingMangaUrlsToId =
-                MangaTable.slice(MangaTable.url, MangaTable.id).select {
+                MangaTable.select {
                     (MangaTable.sourceReference eq sourceId) and (MangaTable.url inList mangas.map { it.url })
-                }.associate { Pair(it[MangaTable.url], it[MangaTable.id].value) }
+                }.associateBy { it[MangaTable.url] }
             val existingMangaUrls = existingMangaUrlsToId.map { it.key }
 
             val mangasToInsert = mangas.filter { !existingMangaUrls.contains(it.url) }
@@ -73,7 +76,42 @@ object MangaList {
             // delete thumbnail in case cached data still exists
             insertedMangaUrlsToId.forEach { (_, id) -> Manga.clearThumbnail(id) }
 
-            val mangaUrlsToId = existingMangaUrlsToId + insertedMangaUrlsToId
+            val mangaToUpdate =
+                mangas.mapNotNull { sManga ->
+                    existingMangaUrlsToId[sManga.url]?.let { sManga to it }
+                }.filterNot { (_, resultRow) ->
+                    resultRow[MangaTable.inLibrary]
+                }
+
+            mangaToUpdate.forEach { (sManga, manga) ->
+                if (manga[MangaTable.thumbnail_url] != sManga.thumbnail_url) {
+                    Manga.clearThumbnail(manga[MangaTable.id].value)
+                }
+            }
+
+            BatchUpdateStatement(MangaTable).apply {
+                mangaToUpdate.forEach { (sManga, manga) ->
+                    addBatch(EntityID(manga[MangaTable.id].value, MangaTable))
+                    this[MangaTable.title] = sManga.title
+                    this[MangaTable.artist] = sManga.artist
+                    this[MangaTable.author] = sManga.author
+                    this[MangaTable.description] = sManga.description
+                    this[MangaTable.genre] = sManga.genre
+                    this[MangaTable.status] = sManga.status
+                    this[MangaTable.thumbnail_url] = sManga.thumbnail_url
+                    this[MangaTable.updateStrategy] = sManga.update_strategy.name
+                    if (!sManga.thumbnail_url.isNullOrEmpty() && manga[MangaTable.thumbnail_url] != sManga.thumbnail_url) {
+                        this[MangaTable.thumbnailUrlLastFetched] = Instant.now().epochSecond
+                    } else {
+                        this[MangaTable.thumbnailUrlLastFetched] = manga[MangaTable.thumbnailUrlLastFetched]
+                    }
+                }
+                execute(this@transaction)
+            }
+
+            val mangaUrlsToId =
+                existingMangaUrlsToId
+                    .mapValues { it.value[MangaTable.id].value } + insertedMangaUrlsToId
 
             mangas.map { manga ->
                 mangaUrlsToId[manga.url]
@@ -86,7 +124,7 @@ object MangaList {
         val mangasPage = this
         val mangaList =
             transaction {
-                val mangaIds = insertOrGet(sourceId)
+                val mangaIds = insertOrUpdate(sourceId)
                 return@transaction MangaTable.select { MangaTable.id inList mangaIds }.map { MangaTable.toDataClass(it) }
             }
         return PagedMangaListDataClass(

--- a/server/src/main/kotlin/suwayomi/tachidesk/manga/impl/MangaList.kt
+++ b/server/src/main/kotlin/suwayomi/tachidesk/manga/impl/MangaList.kt
@@ -83,12 +83,6 @@ object MangaList {
                     resultRow[MangaTable.inLibrary]
                 }
 
-            mangaToUpdate.forEach { (sManga, manga) ->
-                if (manga[MangaTable.thumbnail_url] != sManga.thumbnail_url) {
-                    Manga.clearThumbnail(manga[MangaTable.id].value)
-                }
-            }
-
             BatchUpdateStatement(MangaTable).apply {
                 mangaToUpdate.forEach { (sManga, manga) ->
                     addBatch(EntityID(manga[MangaTable.id].value, MangaTable))
@@ -102,6 +96,7 @@ object MangaList {
                     this[MangaTable.updateStrategy] = sManga.update_strategy.name
                     if (!sManga.thumbnail_url.isNullOrEmpty() && manga[MangaTable.thumbnail_url] != sManga.thumbnail_url) {
                         this[MangaTable.thumbnailUrlLastFetched] = Instant.now().epochSecond
+                        Manga.clearThumbnail(manga[MangaTable.id].value)
                     } else {
                         this[MangaTable.thumbnailUrlLastFetched] = manga[MangaTable.thumbnailUrlLastFetched]
                     }

